doc/explanation/package-management.md

```diff
@@ -0,0 +1,220 @@
+[OPAM-repository](https://github.com/ocaml/opam-repository) which contains all
+the metadata for published packages.
+
+It is managed using Git and OPAM typically uses a snapshot of it to look up the
+dependencies when trying to find a solution that satisfies the constraints.
+
+Likewise, Dune uses the same repository, however instead of snapshots of the
+files it uses the Git repository directly. In fact, Dune maintains a shared
+internal cache containing all Git repositories that projects use. This has the
+advantage that updates to the metadata are very fast since only the newest
+revisions have to be retrieved. The downside is that for the creation of the
+cache the entire repository has to be cloned first.
+
+Given a priorization of fast updates, whenever Dune needs to determine the
+available packages, it will update the repository first. Thus each locking
+process by default will use the newest set of packages available.
+
+However it is also possible to specify specific revisions of the repositories,
+to get a reproducible solution. Due to using Git any previous revision of the
+repository can be used, by specifying a commit hash.
+
+Dune defines two repositories by default:
+
+* `upstream` which refers to the default branch of `opam-repository`, which
+  contains all the publically released packages.
+* `overlay` which refers to
+  [opam-overlay](https://github.com/ocaml-dune/opam-overlays), which defines
+  packages patched to work with package management. The long-term goal is to
+  have as few packages as possible in this repository as more and more packages
+  work within Dune package management upstream. Check the
+  [compatibility](#compatibility) section for details.
+
+### Solving
+
+After Dune has retrieved the constraints and the set of possible packages it is
+necessary to determine which packages & versions should be selected for the
+package lock.
+
+To do so, Dune uses
+[opam-0install-solver](https://github.com/ocaml-opam/opam-0install-solver),
+which is a variant of the 0install solver to find solutions for OPAM packages.
+
+Contrary to `opam`, the Dune solver always starts from a blank slate. It
+assumes nothing is installed and everything needs to be installed. This has the
+advantage that solving is simpler and previous solver solutions don't interfere
+with the current one. Thus given the same inputs it should always have the same
+outputs; no state is held between the solver runs.
+
+This can lead to more packages being installed (as OPAM won't by default
+install new versions of packages if the existing versions satisfy the
+constraints) but it avoids interference from already installed packages that
+lead to potentially different solutions.
+
+After solving is done, the solution gets written into the lock directory with
+all the data that is necessary to build and install the packages. From this
+point, the project contains all necessary information and does not need to
+access any package repositories.
```

Comment:
It might be interesting to have a terminology specification somewhere. When I read the sentence, it seems wrong, as we will later install packages. The distinction between package repositories and the packages we installed should be somewhere in an index, IMO.

Reply:
I am unsure what you consider untrue about it? I've elaborated that section, also noting that it does not download sources, is it better now?

Reply:
Sorry for the delay, as it was marked as outdated I didn't see it. The `package repositories` can be confusing: we won't access opam-repository for the metadata, but it will need to access the server where you host the package to fetch the tarball. It could be considered as a package repository too. Hence, my question about the terminology ^^

Reply:
Ok, I added the word "metadata" (and you can see there is also a note that sources are not being downloaded).

I personally don't consider the package sources a repository per-se as for a repository to me implies a sort of centralized collection of things (like a git repository is a collection of files that belong to a project or opam repository is a collection of opam files or the NPM registry is a repository because it contains the packages themselves), whereas the source files are just a loose collection of URLs. While it's true that a lot of tarballs are on Github, that seems more an emergent outcome than some deliberate effort in creating a repository of package tarballs.

Anyway, long story short, I'm marking this as resolved unless people disagree.